src/handleSubmit.js

```diff
@@ -9,88 +10,86 @@ const handleSubmit = (submit, props, valid, asyncValidate, fields) => {
-            setSubmitSucceeded()
-            if (onSubmitSuccess) {
-              onSubmitSuccess(submitResult, dispatch)
+    })
+    
+    .then(() => {
+      if (asyncValidate) {
+        return Promise.resolve(asyncValidate())
+          .then(asyncErrors => {
+            if (asyncErrors) {
+              asyncValidationFailed = true
+              throw new SubmissionError(asyncErrors)
             }
-            return submitResult
-          }, submitError => {
-            const error = submitError instanceof SubmissionError ? submitError.errors : undefined
-            stopSubmit(error)
-            setSubmitFailed(...fields)
-            if (onSubmitFail) {
-              onSubmitFail(error, dispatch)
-            }
-            if (error || onSubmitFail) {
-              // if you've provided an onSubmitFail callback, don't re-throw the error
-              return error
-            } else {
-              throw submitError
+          }, asyncErrors => {
+            asyncValidationFailed = true
+            if (!(asyncErrors instanceof Error)) {
+              throw new SubmissionError(asyncErrors)
             }
+            throw asyncErrors
           })
-      } else {
-        setSubmitSucceeded()
-        if (onSubmitSuccess) {
-          onSubmitSuccess(result, dispatch)
-        }
       }
+    })
+    
+    .then(() => {
+      const result = submit(values, dispatch, props)
+
+      if (isPromise(result)) {
+        doStartSubmit()
+      }
+      return result
+    })
+    
+    .then(result => {
+      doStopSubmit()
+      setSubmitSucceeded()
+      if (onSubmitSuccess) {
+        onSubmitSuccess(result, dispatch)
+      }
+      
       return result
-    }
+    })
+    
+    .catch(submitError => {
+      const errors = submitError instanceof SubmissionError ? submitError.errors : undefined
+      doStopSubmit(errors)
+      
+      setSubmitFailed(...fields)
 
-    const asyncValidateResult = asyncValidate && asyncValidate()
-    if (asyncValidateResult) {
-      return asyncValidateResult
-        .then(asyncErrors => {
-          if (asyncErrors) {
-            throw asyncErrors
-          }
-          return doSubmit()
-        })
-        .catch(asyncErrors => {
-          setSubmitFailed(...fields)
-          if (onSubmitFail) {
-            onSubmitFail(asyncErrors, dispatch)
-          }
-          return Promise.reject(asyncErrors)
-        })
-    } else {
-      return doSubmit()
-    }
-  } else {
-    setSubmitFailed(...fields)
-    if (onSubmitFail) {
-      onSubmitFail(syncErrors, dispatch)
-    }
-    return syncErrors
-  }
+      if (onSubmitFail) {
+        onSubmitFail(errors, dispatch, submitError)
+      }
+
+      if (errors) {
+        if (asyncValidationFailed) {
+          return Promise.reject(errors)
+        } else {
+          return errors
+        }
+      } else {
+        if (!onSubmitFail) {
+          return Promise.reject(submitError)
+        }
+      }
+    })
```

Comment:
These are the current results:
- If there are errors from asyncValidate: reject the promise with the errors
- If there are errors from validate or SubmissionError: resolve the promise with the errors
- In case of another error: reject the promise with the error, if onSubmitFail was not provided

I think it should be this way:
- If there are errors from asyncValidate, validate or SubmissionError: resolve the promise with the errors
- In case of another error: reject the promise with the error, if onSubmitFail was not provided

The reason for resolving the promise in case of errors is that rejecting would result in unhandled rejection warnings. Also, the errors have been stored in redux by stopSubmit and I would consider that handled.